### PR TITLE
Refactor grammar parser

### PR DESCRIPTION
**Pull request by mfn**
This is the same idea as #30:
- import all classes
- added type declarations were easily possible
- try to un-nest if conditions were possible
- renamed `loadPlugins` to `addPlugins` (that was my fault, no idea wht I came up with `load`…)
- added liberal whitespace in selected parts to make code less dense and more readable

No functional changes intended.

### CODE REVIEW

File: src/GrammarParser.php [hunk 1]
```diff
@@ -73,7 +88,12 @@ static public function getInstance()
         return self::$instance;
     }
 
-    public function buildGrammar($grammarStr, $options = [])
+    /**
+     * @param string $grammarStr
+     * @param array $options
+     * @return Branch[]
```
Comment (mfn) [inline src/GrammarParser.php:94]: This `return` makes it more explicit what to expect (hope I got it right)

File: src/GrammarParser.php [hunk 2]
```diff
@@ -92,12 +112,12 @@ public function buildGrammar($grammarStr, $options = [])
                 $branchName = (string)$grammarBranch->findFirst('branchName');
                 $branchTypeNodeStr = (string)$grammarBranch->findFirst('branchType');
                 $branchType = $branchTypeNodeStr ? substr($branchTypeNodeStr, 1, -1) : $defaultBranchType;
-                $grammar[$branchName] = \ParserGenerator\GrammarNode\BranchFactory::createBranch($branchType,
-                    $branchName);
-            } else {
```
Comment (mfn) [inline src/GrammarParser.php]: Un-nesting

File: src/GrammarParser.php [hunk 3]
```diff
@@ -120,16 +141,17 @@ public function buildGrammar($grammarStr, $options = [])
                 }
 
                 $grammar[$branchName]->setNode($rules);
-            } else {
```
Comment (mfn) [inline src/GrammarParser.php]: Un-nesting

File: src/GrammarParser.php [hunk 4]
```diff
@@ -138,35 +160,34 @@ public function buildGrammar($grammarStr, $options = [])
         return $grammar;
     }
 
-    public function getParser()
+    public function getParser(): Parser
     {
         if ($this->parserSchouldBeRefreshed) {
-            $this->generateNewParser();
+            $this->parser = $this->generateNewParser();
```
Comment (mfn) [inline src/GrammarParser.php:160]: I decided to make the behaviour here explicit so reading the code makes it clear that `->parser` is explicitly set (refreshed) not implicitly in the method.

File: src/GrammarParser.php [hunk 5]
```diff
@@ -138,35 +160,34 @@ public function buildGrammar($grammarStr, $options = [])
...
 
-        $q = new \ParserGenerator\GrammarNode\ItemRestrictions(
-            new \ParserGenerator\GrammarNode\Regex('/[A-Za-z_][0-9A-Za-z_]*/', true),
-            new \ParserGenerator\Extension\ItemRestrictions\ItemRestrictionNot(
-                new \ParserGenerator\Extension\ItemRestrictions\ItemRestrictionOr($restrictedWordsGrammarNode)
+        return
+            new GrammarItemRestrictions(
+                new GrammarRegex('/[A-Za-z_][0-9A-Za-z_]*/', true),
+                new ItemRestrictionNot(
+                    new ItemRestrictionOr($restrictedWordsGrammarNode)
```
Comment (mfn) [inline src/GrammarParser.php:180]: I ❤️ how much more readable this got…

File: src/GrammarParser.php [hunk 6]
```diff
@@ -200,60 +221,56 @@ protected function generateNewParser()
...
             foreach ($rule->findAll('sequenceItem') as $sequenceItem) {
                 $sequenceItemNode = $this->buildSequenceItem($grammar, $sequenceItem, $options);
                 if (count($sequence)) {
                     $sequenceItemNode = new ErrorTrackDecorator($sequenceItemNode);
                 }
                 $sequence[] = $sequenceItemNode;
             }
+
             return $sequence;
-        } else {
```
Comment (mfn) [inline src/GrammarParser.php:214]: Un-nesting

File: src/GrammarParser.php [hunk 7]
```diff
@@ -200,60 +221,56 @@ protected function generateNewParser()
...
         foreach ($this->plugins as $plugin) {
             if ($newSequenceItem = $plugin->buildSequenceItem($grammar, $sequenceItem, $this, $options)) {
-                break;
+                return $newSequenceItem;
             }
         }
 
-        if ($newSequenceItem) {
-            return $newSequenceItem;
-        } else {
```
Comment (mfn) [inline src/GrammarParser.php:242]: Un-nesting

File: src/GrammarParser.php (hunk 2 above)
Comment (farafiri) [reply src/GrammarParser.php]: I don't like it.

I think old fashion 
if (condition) {
  codetrue
} else {
  codefalse
}

is more readable than:

if (condition) {
  codetrue
  continue;
}
codefalse

File: src/GrammarParser.php (hunk 3 above)
Comment (farafiri) [reply src/GrammarParser.php]: same as above

File: src/GrammarParser.php (hunk 5 above)
Comment (farafiri) [reply src/GrammarParser.php:180]: Yep. I have no idea what kind of madnes made me use full class name.

File: src/GrammarParser.php (hunk 2 above)
Comment (mfn) [reply src/GrammarParser.php]: 👍  => undid the change!

File: src/GrammarParser.php (hunk 3 above)
Comment (mfn) [reply src/GrammarParser.php]: 👍  => undid the change!

**Pull request merged**